task4/pkg/api/product.go

```diff
@@ -0,0 +1,121 @@
+package api
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"strconv"
+
+	"github.com/dimaxdqwerty/golang-training-shop/task_4/pkg/data"
+
+	"github.com/gorilla/mux"
+)
+
+type productAPI struct {
+	data *data.ProductData
+}
+
+func ServeProductResource(r *mux.Router, data data.ProductData) {
+	api := &productAPI{data: &data}
+	r.HandleFunc("/products", api.getAllProducts).Methods("GET")
+	r.HandleFunc("/products", api.createUser).Methods("POST")
+	r.HandleFunc("/products/{id}", api.getProduct).Methods("GET")
+	r.HandleFunc("/products/{id}/{unit_price}", api.updateProduct).Methods("PUT")
+	r.HandleFunc("/products/{id}", api.deleteProduct).Methods("DELETE")
+}
+
+func (p productAPI) getAllProducts(writer http.ResponseWriter, request *http.Request) {
+	products, err := p.data.ReadAll()
+	if err != nil {
+		_, err := writer.Write([]byte("got an error when tried to get products"))
+		if err != nil {
+			log.Println(err)
+		}
+	}
```

Comment:
You just printing an error, and not exit from function, so in step: encoding go model to json you will get an error again.
So please add writer.WriteHeader(http.StatusInternalServerError) and exit of function after printing log message.